overlord/devicestate/firstboot_preseed_test.go

```diff
@@ -426,3 +429,211 @@ snaps:
+	chg := st.NewChange("seed", "run the populate from seed changes")
+	for _, ts := range tsAll {
+		chg.AddAll(ts)
+	}
+	c.Assert(st.Changes(), HasLen, 1)
+
+	checkPreseedOrder(c, tsAll, "snapd", "core18", "foo", "bar")
+
+	st.Unlock()
+	err = s.overlord.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+	c.Assert(chg.Err(), IsNil)
+	c.Check(chg.Status(), Equals, state.DoingStatus)
+
+	// we are done with this instance of the overlord, stop it here. Otherwise
+	// it will interfere with our second overlord instance
+	c.Assert(s.overlord.Stop(), IsNil)
+
+	// Verify state between the two change runs
+	r, err := os.Open(dirs.SnapStateFile)
+	c.Assert(err, IsNil)
+	diskState, err := state.ReadState(nil, r)
+	c.Assert(err, IsNil)
+
+	diskState.Lock()
+	defer diskState.Unlock()
+
+	// seeded snaps are installed
+	_, err = snapstate.CurrentInfo(diskState, "snapd")
+	c.Check(err, IsNil)
+	_, err = snapstate.CurrentInfo(diskState, "core18")
+	c.Check(err, IsNil)
+	_, err = snapstate.CurrentInfo(diskState, "foo")
+	c.Check(err, IsNil)
+	_, err = snapstate.CurrentInfo(diskState, "bar")
+	c.Check(err, IsNil)
+
+	// but we're not considered seeded
+	var seeded bool
+	err = diskState.Get("seeded", &seeded)
+	c.Assert(err, testutil.ErrorIs, state.ErrNoState)
+
+	// For the next step of the test, we want to turn off pre-seeding so
+	// we can run the change all the way through, and also see the hooks go
+	// off.
+	restore = snapdenv.MockPreseeding(false)
+	defer restore()
+
+	// Create a new overlord after turning pre-seeding off to run the
+	// change fully through, which we cannot do in pre-seed mode. To actually
+	// invoke the hooks we have to restart the overlord.
+	s.startOverlord(c)
+	st = s.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// avoid device reg
+	chg1 := st.NewChange("become-operational", "init device")
+	chg1.SetStatus(state.DoingStatus)
+
+	st.Unlock()
+	err = s.overlord.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	restart.MockPending(st, restart.RestartUnset)
+	st.Unlock()
+	err = s.overlord.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+	c.Assert(s.overlord.Stop(), IsNil)
+	c.Assert(err, IsNil)
+
+	// Update the change pointer to the change in the new state
+	// otherwise we will be referring to the old one.
+	chg = st.Changes()[0]
+	c.Assert(chg.Err(), IsNil)
+	c.Check(chg.Status(), Equals, state.DoneStatus)
+
+	c.Assert(hooksCalled, HasLen, 1)
+	c.Check(hooksCalled[0].HookName(), Equals, "connect-plug-network")
+
+	// and ensure state is now considered seeded
+	err = st.Get("seeded", &seeded)
+	c.Assert(err, IsNil)
+	c.Check(seeded, Equals, true)
+
+	// check we set seed-time
+	var seedTime time.Time
+	err = st.Get("seed-time", &seedTime)
+	c.Assert(err, IsNil)
+	c.Check(seedTime.IsZero(), Equals, false)
+
+	// verify that connections was made
+	var conns map[string]interface{}
+	c.Assert(st.Get("conns", &conns), IsNil)
+	c.Assert(conns, HasLen, 2)
```

Comment:
nitpick: I'd remove this check and leave only the deep comparison, so that when it fails we actually see the contents.